### PR TITLE
iox-1391 Move `vocabulary types` to separate module

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* Move `optional`, `expected`, `variant` and `string` to their own module `vocabulary types`
  * Remove `cxx` namespace for those four classes
* Move `is_cxx_string` type trait out of the general `type_traits.hpp` into its own file

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1391 (partly)

### CODE REVIEW

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/algorithm.hpp [hunk 1]
```diff
@@ -122,6 +122,8 @@ doesContainValue(const T value, const T firstValueListEntry, const ValueList...
 /// @param[in] v1 the first sorted input container
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/algorithm.hpp:122]: Please add the subscript operator in the line above as well.

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_type_traits.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+//////////////////
+
+/// @brief Provides a translation from a type into its human readable name
+/// NOLINTJUSTIFICATION The name should be stored in a compile time variable. Access is always
+///   safe since it is null terminated and always constant. Other alternatives are not available
+///   at compile time.
+/// NOLINTBEGIN(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+
+template <typename T>
+struct TypeInfo
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/vocabulary/include/iox/detail/string_type_traits.hpp:50]: I'm a bit torn here. I understand that you want to separate the string type traits from others so that a user doesn't depend on the string when not necessary. But it doesn't feel right to have the `TypeInfo` for the integer types here as well.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/algorithm.hpp (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/algorithm.hpp:122]: @FerdinandSpitzschnueffler Not needed anymore as will become private API without doxygen documentation.

File: iceoryx_dust/include/iceoryx_dust/posix_wrapper/named_pipe.hpp [hunk 3]
```diff
@@ -17,13 +17,13 @@
...
 
 #include "iceoryx_hoofs/concurrent/lockfree_queue.hpp"
-#include "iceoryx_hoofs/cxx/string.hpp"
 #include "iceoryx_hoofs/design_pattern/creation.hpp"
 #include "iceoryx_hoofs/internal/posix_wrapper/ipc_channel.hpp"
 #include "iceoryx_hoofs/internal/posix_wrapper/shared_memory_object.hpp"
 #include "iceoryx_hoofs/internal/units/duration.hpp"
 #include "iceoryx_hoofs/posix_wrapper/unnamed_semaphore.hpp"
 #include "iceoryx_platform/semaphore.hpp"
+#include "iox/string.hpp"
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_dust/include/iceoryx_dust/posix_wrapper/named_pipe.hpp:26]: I suppose they are indirectly included. Anyway, we should make that explicit. Added them.

File: iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp [hunk 4]
```diff
@@ -15,12 +15,13 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include "iceoryx_hoofs/cxx/expected.hpp"
-#include "iceoryx_hoofs/cxx/string.hpp"
+#include "iox/expected.hpp"
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp:18]: Good point, done.

File: iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp [hunk 5]
```diff
@@ -124,8 +124,8 @@ class ServiceDescription
 
     /// @brief de-serialization of a ServiceDescription.
     /// @param[in] serialized, Serialization object from which the ServiceDescription shall be created
-    /// @return cxx::expected that either has a ServiceDescription or cxx::Serialization::Error stored inside
-    static cxx::expected<ServiceDescription, cxx::Serialization::Error>
+    /// @return expected that either has a ServiceDescription or cxx::Serialization::Error stored inside
+    static expected<ServiceDescription, cxx::Serialization::Error>
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp:128]: @FerdinandSpitzschnueffler I included `expected`. `optional` is not directly used.

File: iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp [hunk 6]
```diff
@@ -18,10 +18,10 @@
 #define IOX_POSH_CAPRO_SERVICE_DESCRIPTION_HPP
 
 #include "iceoryx_hoofs/cxx/serialization.hpp"
-#include "iceoryx_hoofs/cxx/string.hpp"
 #include "iceoryx_hoofs/cxx/vector.hpp"
 #include "iceoryx_hoofs/log/logstream.hpp"
 #include "iceoryx_posh/iceoryx_posh_types.hpp"
+#include "iox/string.hpp"
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp:24]: @FerdinandSpitzschnueffler Nope, the idea was to adapt `iceoryx_posh` right away. However, for other projects who are still using the old include path we provide the legacy folder. This legacy header will become deprecated shortly before the C-release `v3.0`.

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_type_traits.hpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/string_type_traits.hpp:50]: @FerdinandSpitzschnueffler I understand what you mean. However, I came to this decision after looking at the usage of the header `string_type_traits.hpp`:

1. `functional_interface.inl`
2. `string.hpp`

`TypeInfo` is sort of related to string as well as it `Provides a translation from a type into its human readable name`. Happy to rename the file or split it up into even smaller chunks. But I suppose that's something for a follow PR as this one is already quite big.

**Comment by elBoberido** [comment]
Please fix the links to the `expected` and `optional` in `how-optional-and-error-values-are-returned-in-iceoryx.md`

**Comment by mossmaurice** [comment]
@elBoberido 

> Please fix the links to the `expected` and `optional` in `how-optional-and-error-values-are-returned-in-iceoryx.md`

Good catch, done :heavy_check_mark:

File: iceoryx_hoofs/vocabulary/include/iox/expected.hpp [hunk 7]
```diff
@@ -14,27 +14,32 @@
...
 
 namespace iox
 {
-namespace cxx
+namespace internal
 {
+/// @brief Type trait which verifies whether the passed type T is of type optional
+template <typename...>
+struct IsOptional;
+} // namespace internal
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/vocabulary/include/iox/expected.hpp:35]: Why is this needed?

Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/expected.hpp:35]: Good catch. I think in the meantime this was removed or renamed to `is_optional` on `master`. Removed it.

File: iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl [hunk 8]
```diff
@@ -14,17 +14,27 @@
...
+template <typename... T>
+struct IsOptional : std::false_type
+{
+};
+template <typename T>
+struct IsOptional<iox::optional<T>> : std::true_type
+{
+};
+} // namespace internal
+
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl:37]: I removed the above code.

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_type_traits.hpp (hunk 2 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/vocabulary/include/iox/detail/string_type_traits.hpp:50]: Fine for now, I guess. I don't see an elegant way to separate `TypeInfo` for integer types and string at the moment.

File: iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl (hunk 8 above)
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl]: @FerdinandSpitzschnueffler Because in the meantime `IsOptional` was renamed to `is_optional`, see [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1791#discussion_r1066275925).

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+#include "iox/expected.hpp"
+
+namespace iox
+{
+/// @todo iox-#1593 Deprecate include
+/// [[deprecated("Deprecated in 3.0, removed in 4.0, please include 'iox/expected.hpp' instead")]]
+namespace cxx
+{
+/// @deprecated
+using iox::expected;
```
Comment (elBoberido) [inline iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:29]: Shouldn't `iox::success` and `iox::error` also be imported to the deprecated namespace?

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "iox/expected.hpp"
+
+namespace iox
+{
+/// @todo iox-#1593 Deprecate include
+/// [[deprecated("Deprecated in 3.0, removed in 4.0, please include 'iox/expected.hpp' instead")]]
+namespace cxx
+{
+/// @deprecated
```
Comment (elBoberido) [inline iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:28]:
```suggestion
/// @deprecated use `iox::expected` instead of `iox::cxx::expected`
```

This gives more context

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
Looks good but please update `./clang-tidy-diff-scans.txt`.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elBoberido — COMMENTED** [review]
Second half done :)

File: doc/design/error-handling.md [hunk 11]
```diff
@@ -54,7 +54,7 @@ To this end it is necessary to eliminate the use of all potentially throwing (th
 
 #### Alternatives to exceptions
 
-As an alternative to exceptions we use the error handler and a variation of return codes in the form of ``cxx::expected``, described below. ``cxx::expected`` can be used to communicate the error to the caller, who has to decide whether to handle the error itself or propagate it further (e.g. as another ``cxx::expected``). Error handling itself is performed by the error handler which handles errors occurring in the subcomponents of ``iceoryx::posh``.
+As an alternative to exceptions we use the error handler and a variation of return codes in the form of ``iox::expected``, described below. ``iox::expected`` can be used to communicate the error to the caller, who has to decide whether to handle the error itself or propagate it further (e.g. as another ``iox::expected``). Error handling itself is performed by the error handler which handles errors occurring in the subcomponents of ``iceoryx::posh``.
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply doc/design/error-handling.md:57]: :heavy_check_mark: It's now `iox::foo` everywhere.

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/string.hpp [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+#include "iox/string.hpp"
+
+namespace iox
+{
+/// @todo iox-#1593 Deprecate include
+/// [[deprecated("Deprecated in 3.0, removed in 4.0, please include 'iox/string.hpp' instead")]]
+namespace cxx
+{
+/// @deprecated
+using iox::string;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/string.hpp:29]: @elBoberido I added and will build against the old posh to find any missing API.

**Comment by mossmaurice** [comment]
@elBoberido Thanks for the hint!

> Please do a git checkout master iceoryx_posh and make it compile. This ensures that all symbols are available via the legacy includes.

 This revealed all the places in `iceoryx_posh` where we implicitly including one of those files  :grimacing: :grin:  I'll fix those on a separate PR.

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+#ifndef IOX_HOOFS_CXX_EXPECTED_HPP
+#define IOX_HOOFS_CXX_EXPECTED_HPP
+
+#include "iox/expected.hpp"
+
+namespace iox
+{
+/// @todo iox-#1593 Deprecate include
+/// [[deprecated("Deprecated in 3.0, removed in 4.0, please include 'iox/expected.hpp' instead")]]
```
Comment (elBoberido) [inline iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:25]: I'm wondering a bit when this deprecation warning would be shown. When the `cxx` namespace will be used? In this case the deprecation message would be wrong

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+namespace cxx
+{
+/// @deprecated use `iox::expected` instead of `iox::cxx::expected`
+using iox::error;
+using iox::expected;
```
Comment (elBoberido) [inline iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:30]: Please add the deprecated documentation to all using directives

**Review by elBoberido — COMMENTED** [review]
Looks good. Almost there

**Comment by elBoberido** [comment]
Thinking about it, we could use this trick to ensure that we do not break the API. A CI job which builds posh 3.0 against hoofs master and examples 3.0 against posh + hoofs master.

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp (hunk 13 above)
Comment (mossmaurice) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:25]: @elBoberido I tried this and had a look very early on when we started with #1391. This will only deprecated what is in exactly this very namespace definition and not everything in `cxx` :D

Comment (elBoberido) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:25]: Okay, tried it out. Deprecating namespaces is a C++17 feature and would indeed print it everywhere where the namespace is used. I think we have to add this comment to the respective using directive :(

**Comment by mossmaurice** [comment]
@elBoberido 

> Thinking about it, we could use this trick to ensure that we do not break the API. A CI job which builds posh 3.0 against hoofs master and examples 3.0 against posh + hoofs master.

IMHO that would be overkill...

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp (hunk 13 above)
Comment (mossmaurice) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:25]: @elBoberido Odd, I tried that :thinking: According to cppref it's C++14: https://en.cppreference.com/w/cpp/language/attributes/deprecated

**Comment by elBoberido** [comment]
> IMHO that would be overkill...

well, it depends on how serious we are about API breakage :sweat_smile: 

we might not yet be there  but at some point I think it would make sense

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp (hunk 13 above)
Comment (mossmaurice) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:25]: Ah, I might know why it didn't work for you. IIRC you have to write:

```
namespace [[deprecated]] foo
{
  // bla..
}
```

Comment (elBoberido) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:25]: I get this warning from my IDE `Attributes on a namespace declaration are a C++17 extension [-Wc++17-extensions]` and when compiling with clang but not with gcc

**Review by elBoberido — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
